tests/loops/test_loops.py

```diff
@@ -123,99 +111,375 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
+
+    model = ValidationModel()
+    model.validation_epoch_end = None
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=n_epochs,
+        limit_train_batches=1,
+        limit_val_batches=n_batches,
+        num_sanity_val_steps=0,
+    )
+
+    # simulate a failure
+    try:
+        trainer.fit(model)
+    except CustomException:
+        pass
+
+    ckpt_path = str(tmpdir / '.pl_auto_save.ckpt')
+    checkpoint = torch.load(ckpt_path)["loops"]["fit_loop"]
+
+    total_dataloader = stop_epoch * n_dataloaders + stop_dataloader
+    expected = {
+        "total": {
+            "ready": total_dataloader + 1,
+            "started": None,
+            "processed": None,
+            "completed": total_dataloader
+        },
+        "current": {
+            "ready": stop_dataloader + 1,
+            "started": None,
+            "processed": None,
+            "completed": stop_dataloader,
+        },
+    }
+    assert checkpoint["epoch_loop.val_loop.dataloader_progress"] == expected
+
+    trainer.fit_loop.load_state_dict(checkpoint, restart_progress=False)
+
+    # `nbe_`: non-breaking epoch, as in, no exception will be raised. `be_`: breaking epoch
+    nbe_total_val_batch = stop_epoch * n_dataloaders * n_batches
+    be_total_val_batch = stop_dataloader * n_batches + stop_batch
+    total_val_batch = nbe_total_val_batch + be_total_val_batch
+    expected = {
+        "total": {
+            "ready": total_val_batch + 1,
+            "started": total_val_batch + 1,
+            "processed": total_val_batch,
+            "completed": total_val_batch
+        },
+        "current": {
+            "ready": stop_batch + 1,
+            "started": stop_batch + 1,
+            "processed": stop_batch,
+            "completed": stop_batch,
+        },
+    }
+    assert trainer.fit_loop.epoch_loop.val_loop.epoch_loop.batch_progress.state_dict() == expected
+
+    trainer.fit_loop.load_state_dict(checkpoint)
+    expected = {
+        "total": {
+            "ready": total_val_batch,
+            "started": total_val_batch,
+            "processed": total_val_batch,
+            "completed": total_val_batch
+        },
+        "current": {
+            "ready": stop_batch,
+            "started": stop_batch,
+            "processed": stop_batch,
+            "completed": stop_batch
+        },
+    }
+    assert trainer.fit_loop.epoch_loop.val_loop.epoch_loop.batch_progress.state_dict() == expected
+
+
+@mock.patch.dict(os.environ, {"PL_FAULT_TOLERANT_TRAINING": "1"})
+@pytest.mark.parametrize("accumulate_grad_batches", (1, 2, 3))
+@pytest.mark.parametrize("n_optimizers", (1, 3, 5))
+@pytest.mark.parametrize("stop_epoch", (1, 2))
+@pytest.mark.parametrize("stop_batch", (1, 2))
+@pytest.mark.parametrize("stop_optimizer", (1, 2))
+def test_loop_state_on_exception(accumulate_grad_batches, stop_epoch, stop_batch, stop_optimizer, n_optimizers, tmpdir):
```

Comment:
Just for anyone coming to this PR, @carmocca pointed me to these tests which gave me a lot of confidence that the behaviour is as expected. Very strong tests to ensure that the behaviour is correct!